### PR TITLE
[JBPM-9129] Start process without triggering a start node

### PR DESCRIPTION
**Pull request by elguardian**
Added functions to trigger a new process from an arbitrary node.

### CODE REVIEW

**Comment by elguardian** [comment]
depends on https://github.com/kiegroup/droolsjbpm-knowledge/pull/438

**Comment by elguardian** [comment]
Jekins restest this please

**Comment by elguardian** [comment]
this is ok to test

**Comment by elguardian** [comment]
Jenkins this is ok to test

**Comment by elguardian** [comment]
Jenkins retest this please

**Comment by elguardian** [comment]
Jenkins restest this please.

**Comment by elguardian** [comment]
Jenkins retest this please.

File: jbpm-flow/src/main/java/org/jbpm/process/instance/impl/CancelNodeInstanceAction.java [hunk 1]
```diff
@@ -38,7 +40,7 @@ public void execute(ProcessContext context) throws Exception {
 		WorkflowProcessInstance pi = context.getNodeInstance().getProcessInstance();
 		NodeInstance nodeInstance = findNodeByUniqueId(pi.getNodeInstances(), attachedToNodeId);
 		if (nodeInstance != null) {
-		    ((org.jbpm.workflow.instance.NodeInstance)nodeInstance).cancel();
+            ((org.jbpm.workflow.instance.NodeInstance) nodeInstance).cancel(TIMED);
```
Comment (krisv) [inline jbpm-flow/src/main/java/org/jbpm/process/instance/impl/CancelNodeInstanceAction.java:43]: I think CancelNodeInstanceAction is used in multiple cases, boundary escalation and boundary error events, not sure I grasp the meaning of TIMED cancel type in this context?

File: jbpm-flow/src/main/java/org/jbpm/ruleflow/instance/RuleFlowProcessInstance.java [hunk 2]
```diff
@@ -46,4 +48,32 @@ public void internalStart(String trigger) {
     	    .forEach(austoStartNode -> signalEvent(austoStartNode.getName(), null));
     }
 
+
+    private void trigger(RuleFlowProcessInstance wfp, String nodeId) {
```
Comment (krisv) [inline jbpm-flow/src/main/java/org/jbpm/ruleflow/instance/RuleFlowProcessInstance.java:52]: What's the point of passing wfp here, isn't this always this?

File: jbpm-flow/src/main/java/org/jbpm/workflow/instance/impl/NodeInstanceImpl.java [hunk 3]
```diff
@@ -78,12 +82,12 @@
     protected int slaCompliance = ProcessInstance.SLA_NA;
     protected Date slaDueDate;
     protected long slaTimerId = -1;
-    protected boolean aborted = false;
```
Comment (krisv) [inline jbpm-flow/src/main/java/org/jbpm/workflow/instance/impl/NodeInstanceImpl.java:81]: Node instances are stored as part of the binary state using protobuf, wouldn't we have to keep the aborted field to keep backwards compatibility?  Would we need to generate the protobuf files?  From what I understand this field is just used to pass information around (to the listener generating the logs), so I guess it probably isn't going to be necessary to be serialized (actually aborted node instances should not be serialized), but unless this aborted field was somehow excluded in protobuf marshalling, and/or cancelType will be too?

File: jbpm-services/jbpm-kie-services/src/main/java/org/jbpm/kie/services/impl/RuntimeDataServiceImpl.java [hunk 4]
```diff
@@ -761,6 +761,13 @@ public NodeInstanceDesc getNodeInstanceForWorkItem(Long workItemId) {
         return variablesState;
     }
 
+    @Override
+    public Collection<VariableDesc> getVariablesCurrentStateWithValueObject(Long processInstanceId) {
```
Comment (krisv) [inline jbpm-services/jbpm-kie-services/src/main/java/org/jbpm/kie/services/impl/RuntimeDataServiceImpl.java:765]: It seems most of the logic around storying actual object was removed for more review, should this (and similar references in this PR to WithValueObject) still be here?

File: jbpm-flow/src/main/java/org/jbpm/process/instance/impl/CancelNodeInstanceAction.java (hunk 1 above)
Comment (elguardian) [reply jbpm-flow/src/main/java/org/jbpm/process/instance/impl/CancelNodeInstanceAction.java:43]: You are right. I can use SKIPPED vs OBSOLETE. OBSOLETE means the task has no meaning anymore as the group it belonged is already completted. SKIPPED would mean that another route has been taking.

File: jbpm-flow/src/main/java/org/jbpm/ruleflow/instance/RuleFlowProcessInstance.java (hunk 2 above)
Comment (elguardian) [reply jbpm-flow/src/main/java/org/jbpm/ruleflow/instance/RuleFlowProcessInstance.java:52]: sure

File: jbpm-flow/src/main/java/org/jbpm/workflow/instance/impl/NodeInstanceImpl.java (hunk 3 above)
Comment (elguardian) [reply jbpm-flow/src/main/java/org/jbpm/workflow/instance/impl/NodeInstanceImpl.java:81]: made it transient.

File: jbpm-services/jbpm-kie-services/src/main/java/org/jbpm/kie/services/impl/RuntimeDataServiceImpl.java (hunk 4 above)
Comment (elguardian) [reply jbpm-services/jbpm-kie-services/src/main/java/org/jbpm/kie/services/impl/RuntimeDataServiceImpl.java:765]: removed. left behind by mistake.

**Comment by elguardian** [comment]
done @krisv

**Comment by elguardian** [comment]
Jenkins retest this please.

File: jbpm-test-util/src/main/java/org/jbpm/test/util/Version.java [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+    @Override
+    public int compareTo(Version other) {
+        if (major.equals(other.major)) {
+            return minor.compareTo(other.minor);
+        }
+        return major.compareTo(other.major);
+    }
+
+    @Override
+    public boolean equals(Object obj) {
```
Comment (afalhambra) [inline jbpm-test-util/src/main/java/org/jbpm/test/util/Version.java:48]: hasCode method should be implemented, but up to you as this is a test class.

File: jbpm-test-util/src/main/java/org/jbpm/test/util/DatabaseScript.java [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+                return this.name.compareTo(o.name);
+            }
+            return from.compareTo(o.from);
+        } else {
+            return this.name.compareTo(o.name);
+        }
+    }
+
+    @Override
+    public boolean equals(Object obj) {
```
Comment (afalhambra) [inline jbpm-test-util/src/main/java/org/jbpm/test/util/DatabaseScript.java:78]: hasCode method should be implemented, but up to you as this is a test class.

File: jbpm-flow/src/main/java/org/jbpm/workflow/instance/impl/NodeInstanceImpl.java [hunk 7]
```diff
@@ -149,9 +155,13 @@ public Node getNode() {
     public boolean isInversionOfControl() {
         return false;
     }
+
+    public final void cancel() {
+        cancel(ABORTED);
```
[the remark this answers is not in the document]
Comment (krisv) [reply jbpm-flow/src/main/java/org/jbpm/workflow/instance/impl/NodeInstanceImpl.java:160]: Field aborted and cancelType aren't necessary to be persisted long-term, but more short-term for data passing.  Probably aborted could have been transient as well, but for backwards compability we probably can't change that now.  So I guess the plan is to just keep the field, nothing more.  I think we have tests that are run as part of QE suite that test portability of process instance binary state across versions, can we make sure these are run to make sure we don't break backwards compatibility with this?

[the remark this answers is not in the document]
Comment (afalhambra) [reply jbpm-flow/src/main/java/org/jbpm/workflow/instance/impl/NodeInstanceImpl.java:160]: Yes, there are tests as part of a test suite in `jbpm-runtime-manager` module. However these are related to a migration between different bpm processes. So I'm wondering whether this will be enough or we should add a specific test case for backward compatibility. @krisv, @elguardian wdyt?

https://github.com/kiegroup/jbpm/blob/master/jbpm-runtime-manager/src/test/java/org/jbpm/runtime/manager/impl/migration/MigrationManagerTest.java

[the remark this answers is not in the document]
Comment (krisv) [reply jbpm-flow/src/main/java/org/jbpm/workflow/instance/impl/NodeInstanceImpl.java:160]: This aborted field would be persisted with any node instance, so I think any of those migration tests would be able to detect migration issues caused by this.  There should be sufficient other unit tests for simply testing whether aborting a node instance works well (using cancelType).

File: jbpm-human-task/jbpm-human-task-workitems/src/test/java/org/jbpm/services/task/test/TestStatefulKnowledgeSession.java [hunk 8]
```diff
@@ -312,4 +312,11 @@ public void delete(FactHandle fh, FactHandle.State fhState) {
     public void destroy() {
         dispose();        
     }
+
+    @Override
+    public ProcessInstance startProcessFromNodeIds(String processId, Map<String, Object> params, String... nodeId) {
+        // TODO Auto-generated method stub
```
[the remark this answers is not in the document]
Comment (elguardian) [reply jbpm-human-task/jbpm-human-task-workitems/src/test/java/org/jbpm/services/task/test/TestStatefulKnowledgeSession.java:318]: done

File: jbpm-flow/src/main/java/org/jbpm/ruleflow/instance/RuleFlowProcessInstance.java [hunk 9]
```diff
@@ -46,4 +48,31 @@ public void internalStart(String trigger) {
     	    .forEach(austoStartNode -> signalEvent(austoStartNode.getName(), null));
     }
 
+
+    public void startProcessFromNodeIds(String[] nodeInstancesIds) {
```
[the remark this answers is not in the document]
Comment (elguardian) [reply jbpm-flow/src/main/java/org/jbpm/ruleflow/instance/RuleFlowProcessInstance.java:52]: done

File: jbpm-flow/src/main/java/org/jbpm/workflow/core/impl/NodeContainerImpl.java [hunk 10]
```diff
@@ -63,6 +64,12 @@ public Node getNode(final long id) {
         return node; 
     }
 
+    @Override
+    public Node getNodeByUniqueId(String nodeId) {
+        Optional<Node> node = this.nodes.values().stream().filter(e -> e.getNodeUniqueId().equals(nodeId)).findFirst();
+        return node.isPresent() ? node.get() : null;
```
[the remark this answers is not in the document]
Comment (elguardian) [reply jbpm-flow/src/main/java/org/jbpm/workflow/core/impl/NodeContainerImpl.java:70]: ok

File: jbpm-services/jbpm-kie-services/src/test/java/org/jbpm/kie/services/test/StartProcessFromArbitraryNodesServiceImplTest.java [hunk 11]
```diff
@@ -0,0 +1,225 @@
...
+
+        Thread.sleep(3000);
+
+        processService.abortProcessInstance(processInstanceId);
+
+        Collection<NodeInstanceDesc> timed = runtimeDataService.getProcessInstanceFullHistoryByType(processInstanceId, EntryType.SKIPPED, new QueryContext(0, 0));
+        assertThat(timed.size(), is(1));
+        Collection<NodeInstanceDesc> aborted = runtimeDataService.getProcessInstanceFullHistoryByType(processInstanceId, EntryType.ABORTED, new QueryContext(0, 0));
+        assertThat(aborted.size(), is(1));
+        String[] nodeIds = aborted.stream().map(e -> e.getNodeId()).toArray(String[]::new);
```
[the remark this answers is not in the document]
Comment (elguardian) [reply jbpm-services/jbpm-kie-services/src/test/java/org/jbpm/kie/services/test/StartProcessFromArbitraryNodesServiceImplTest.java:174]: done

File: jbpm-test-util/src/main/java/org/jbpm/test/util/DatabaseScript.java [hunk 12]
```diff
@@ -0,0 +1,89 @@
...
+        if (obj instanceof DatabaseScript) {
+            return this.compareTo((DatabaseScript) obj) == 0;
+        }
+        return false;
+    }
+
+    @Override
+    public String toString() {
+        return "Script " + this.name + " from version " + from + " to " + to;
```
[the remark this answers is not in the document]
Comment (elguardian) [reply jbpm-test-util/src/main/java/org/jbpm/test/util/DatabaseScript.java:87]: done

File: jbpm-test-util/src/main/java/org/jbpm/test/util/DatabaseScript.java (hunk 6 above)
Comment (elguardian) [reply jbpm-test-util/src/main/java/org/jbpm/test/util/DatabaseScript.java:78]: done

File: jbpm-test-util/src/main/java/org/jbpm/test/util/Version.java (hunk 5 above)
Comment (elguardian) [reply jbpm-test-util/src/main/java/org/jbpm/test/util/Version.java:48]: done

File: jbpm-services/jbpm-kie-services/src/test/java/org/jbpm/kie/services/test/StartProcessFromArbitraryNodesServiceImplTest.java [hunk 13]
```diff
@@ -0,0 +1,225 @@
...
+
+
+        Long processInstanceId = processService.startProcess(deploymentUnit.getIdentifier(), processId, singletonMap("var_a", 3));
+        assertNotNull(processInstanceId);
+
+        Thread.sleep(3000);
+
+        processService.abortProcessInstance(processInstanceId);
+
+        Collection<NodeInstanceDesc> timed = runtimeDataService.getProcessInstanceFullHistoryByType(processInstanceId, EntryType.SKIPPED, new QueryContext(0, 0));
```
[the remark this answers is not in the document]
Comment (elguardian) [reply jbpm-services/jbpm-kie-services/src/test/java/org/jbpm/kie/services/test/StartProcessFromArbitraryNodesServiceImplTest.java:170]: done

File: jbpm-services/jbpm-kie-services/src/test/java/org/jbpm/kie/services/test/StartProcessFromArbitraryNodesServiceImplTest.java [hunk 14]
```diff
@@ -0,0 +1,225 @@
...
+    @Before
+    public void prepare() {
+    	configureServices();
+    	logger.debug("Preparing kjar");
+        KieServices ks = KieServices.Factory.get();
+        ReleaseId releaseId = ks.newReleaseId(GROUP_ID, ARTIFACT_ID, VERSION);
+        List<String> processes = new ArrayList<String>();
+        processes.add("repo/processes/general/BPMN2-SimpleRestartTest.bpmn2");
+        processes.add("repo/processes/general/BPMN2-SimpleRestartWithTimerTest.bpmn2");
+        processes.add("repo/processes/general/BPMN2-SimpleRestartWithObsoleteTest.bpmn2");
```
[the remark this answers is not in the document]
Comment (elguardian) [reply jbpm-services/jbpm-kie-services/src/test/java/org/jbpm/kie/services/test/StartProcessFromArbitraryNodesServiceImplTest.java:66]: not relevant

File: jbpm-services/jbpm-kie-services/src/test/java/org/jbpm/kie/services/test/StartProcessFromArbitraryNodesServiceImplTest.java [hunk 15]
```diff
@@ -0,0 +1,225 @@
...
+            FileOutputStream fs = new FileOutputStream(pom);
+            fs.write(getPom(releaseId).getBytes());
+            fs.close();
+        } catch (Exception e) {
+
+        }
+		KieMavenRepository repository = getKieMavenRepository();
+        repository.deployArtifact(releaseId, kJar1, pom);
+
+        ReleaseId releaseId3 = ks.newReleaseId(GROUP_ID, ARTIFACT_ID, "1.1.0");
```
[the remark this answers is not in the document]
Comment (elguardian) [reply jbpm-services/jbpm-kie-services/src/test/java/org/jbpm/kie/services/test/StartProcessFromArbitraryNodesServiceImplTest.java:81]: ok

File: jbpm-services/jbpm-kie-services/src/test/java/org/jbpm/kie/services/test/StartProcessFromArbitraryNodesServiceImplTest.java [hunk 16]
```diff
@@ -0,0 +1,225 @@
...
+        Long processInstanceId = processService.startProcess(deploymentUnit.getIdentifier(), "restart.simple", singletonMap("var_a", 3));
+    	assertNotNull(processInstanceId);
+
+        processService.abortProcessInstance(processInstanceId);
+
+        Collection<NodeInstanceDesc> logs = runtimeDataService.getProcessInstanceFullHistoryByType(processInstanceId, EntryType.ABORTED, new QueryContext(0, 0));
+        String[] nodeIds = logs.stream().map(e -> e.getNodeId()).toArray(String[]::new);
+
+        processInstanceId = processService.startProcessFromNodeIds(deploymentUnit.getIdentifier(), "restart.simple", singletonMap("var_a", 3), nodeIds);
+        runtimeDataService.getTasksByProcessInstanceId(processInstanceId).forEach(e -> {
```
[the remark this answers is not in the document]
Comment (elguardian) [reply jbpm-services/jbpm-kie-services/src/test/java/org/jbpm/kie/services/test/StartProcessFromArbitraryNodesServiceImplTest.java:132]: the goal of the test is to see if the starts node instance work. not really checking the task works or not.

File: jbpm-services/jbpm-kie-services/src/test/java/org/jbpm/kie/services/test/StartProcessFromArbitraryNodesServiceImplTest.java [hunk 17]
```diff
@@ -0,0 +1,225 @@
...
+            this.userTaskService.start(e, "john");
+            this.userTaskService.complete(e, "john", emptyMap());
+        });
+
+        processService.abortProcessInstance(processInstanceId);
+
+        logs = runtimeDataService.getProcessInstanceFullHistoryByType(processInstanceId, EntryType.ABORTED, new QueryContext(0, 0));
```
[the remark this answers is not in the document]
Comment (elguardian) [reply jbpm-services/jbpm-kie-services/src/test/java/org/jbpm/kie/services/test/StartProcessFromArbitraryNodesServiceImplTest.java:139]: the goal of the test is to see if the starts node instance work. not really checking the task works or not.

File: jbpm-services/jbpm-kie-services/src/test/java/org/jbpm/kie/services/test/StartProcessFromArbitraryNodesServiceImplTest.java [hunk 18]
```diff
@@ -0,0 +1,225 @@
...
+        nodeIds = logs.stream().map(e -> e.getNodeId()).toArray(String[]::new);
```
[the remark this answers is not in the document]
Comment (elguardian) [reply jbpm-services/jbpm-kie-services/src/test/java/org/jbpm/kie/services/test/StartProcessFromArbitraryNodesServiceImplTest.java:140]: done

File: jbpm-services/jbpm-kie-services/src/test/java/org/jbpm/kie/services/test/StartProcessFromArbitraryNodesServiceImplTest.java [hunk 19]
```diff
@@ -0,0 +1,225 @@
...
+        processInstanceId = processService.startProcessFromNodeIds(deploymentUnit.getIdentifier(), "restart.simple", singletonMap("var_a", 3), nodeIds);
```
[the remark this answers is not in the document]
Comment (elguardian) [reply jbpm-services/jbpm-kie-services/src/test/java/org/jbpm/kie/services/test/StartProcessFromArbitraryNodesServiceImplTest.java:141]: there is no previous state.

File: jbpm-services/jbpm-kie-services/src/test/java/org/jbpm/kie/services/test/StartProcessFromArbitraryNodesServiceImplTest.java [hunk 20]
```diff
@@ -0,0 +1,225 @@
...
+        runtimeDataService.getTasksByProcessInstanceId(processInstanceId).forEach(e -> {
```
[the remark this answers is not in the document]
Comment (elguardian) [reply jbpm-services/jbpm-kie-services/src/test/java/org/jbpm/kie/services/test/StartProcessFromArbitraryNodesServiceImplTest.java:142]: not the goal of the test.

File: jbpm-services/jbpm-kie-services/src/test/java/org/jbpm/kie/services/test/StartProcessFromArbitraryNodesServiceImplTest.java [hunk 21]
```diff
@@ -0,0 +1,225 @@
...
+
+        deploymentService.deploy(deploymentUnit);
+        units.add(deploymentUnit);
+        assertNotNull(processService);
+
+
+        Long processInstanceId = processService.startProcess(deploymentUnit.getIdentifier(), processId, singletonMap("var_a", 3));
+        assertNotNull(processInstanceId);
+
+        Thread.sleep(3000);
```
[the remark this answers is not in the document]
Comment (elguardian) [reply jbpm-services/jbpm-kie-services/src/test/java/org/jbpm/kie/services/test/StartProcessFromArbitraryNodesServiceImplTest.java:166]: done

File: jbpm-services/jbpm-services-api/src/build/revapi-config.json [hunk 22]
```diff
@@ -71,6 +71,42 @@
...
+                },
+                {
+                    "code": "java.method.addedToInterface",
+                    "new": "method java.lang.Object org.jbpm.services.api.model.VariableDesc::getNewValueObject()",
+                    "package": "org.jbpm.services.api.model",
+                    "classSimpleName": "VariableDesc",
+                    "methodName": "getNewValueObject",
+                    "elementKind": "method",
+                    "justification": "added for retrieving the original object value and not the string representation"
+                },
```
[the remark this answers is not in the document]
Comment (elguardian) [reply jbpm-services/jbpm-services-api/src/build/revapi-config.json:92]: done

File: jbpm-services/jbpm-services-api/src/main/java/org/jbpm/services/api/ProcessService.java [hunk 23]
```diff
@@ -81,12 +81,39 @@
...
+     * 
+     * @param deploymentId deployment information for the process's kjar
+     * @param processId The process's identifier
+     * @param params process variables
+     * @return process instance identifier
+     * @throws RuntimeException in case of encountered errors
+     * @throws DeploymentNotFoundException in case deployment with given deployment id does not exist
+     * @throws DeploymentNotActiveException in case deployment with given deployment id is not active
+     */
+    Long startProcessFromNodeIds(String deploymentId, String processId, Map<String, Object> params, String... nodeIds);
```
[the remark this answers is not in the document]
Comment (elguardian) [reply jbpm-services/jbpm-services-api/src/main/java/org/jbpm/services/api/ProcessService.java:94]: done

File: jbpm-services/jbpm-services-api/src/main/java/org/jbpm/services/api/ProcessService.java [hunk 24]
```diff
@@ -81,12 +81,39 @@
...
+     * @param deploymentId deployment information for the process's kjar
+     * @param processId The process's identifier
+     * @param key correlation key (must be unique)
+     * @param params process variables
+     * @return process instance identifier
+     * @throws RuntimeException in case of encountered errors
+     * @throws DeploymentNotFoundException in case deployment with given deployment id does not exist
+     * @throws DeploymentNotActiveException in case deployment with given deployment id is not active
+     */
+    Long startProcessFromNodeIds(String deploymentId, String processId, CorrelationKey key, Map<String, Object> params, String... nodeIds);
```
[the remark this answers is not in the document]
Comment (elguardian) [reply jbpm-services/jbpm-services-api/src/main/java/org/jbpm/services/api/ProcessService.java:108]: done

File: jbpm-services/jbpm-kie-services/src/test/java/org/jbpm/kie/services/test/StartProcessFromArbitraryNodesServiceImplTest.java [hunk 25]
```diff
@@ -0,0 +1,225 @@
...
+        deploymentService.deploy(deploymentUnit);
+        units.add(deploymentUnit);
+    	assertNotNull(processService);
+
+        Long processInstanceId = processService.startProcess(deploymentUnit.getIdentifier(), "restart.simple", singletonMap("var_a", 3));
+    	assertNotNull(processInstanceId);
+
+        processService.abortProcessInstance(processInstanceId);
+
+        Collection<NodeInstanceDesc> logs = runtimeDataService.getProcessInstanceFullHistoryByType(processInstanceId, EntryType.ABORTED, new QueryContext(0, 0));
```
[the remark this answers is not in the document]
Comment (elguardian) [reply jbpm-services/jbpm-kie-services/src/test/java/org/jbpm/kie/services/test/StartProcessFromArbitraryNodesServiceImplTest.java:128]: done

File: jbpm-services/jbpm-kie-services/src/test/resources/repo/processes/general/BPMN2-SimpleRestartTest.bpmn2 [hunk 26]
```diff
@@ -0,0 +1,351 @@
...
+        <bpmn2:assignment id="_UVLvtorNEeqCP7d-9hkcyw">
+          <bpmn2:from xsi:type="bpmn2:tFormalExpression" id="_UVLvt4rNEeqCP7d-9hkcyw"><![CDATA[false]]></bpmn2:from>
+          <bpmn2:to xsi:type="bpmn2:tFormalExpression" id="_UVLvuIrNEeqCP7d-9hkcyw">_50FE3D26-B167-4A73-BDB9-C55F2A98AE38_SkippableInputX</bpmn2:to>
+        </bpmn2:assignment>
+      </bpmn2:dataInputAssociation>
+      <bpmn2:dataOutputAssociation id="_UVLvuYrNEeqCP7d-9hkcyw">
+        <bpmn2:sourceRef>_50FE3D26-B167-4A73-BDB9-C55F2A98AE38_ht_outOutputX</bpmn2:sourceRef>
+        <bpmn2:targetRef>var_a</bpmn2:targetRef>
+      </bpmn2:dataOutputAssociation>
+      <bpmn2:potentialOwner id="_sjFE3NZxEeWhSJ3J-UY8fA">
```
[the remark this answers is not in the document]
Comment (elguardian) [reply jbpm-services/jbpm-kie-services/src/test/resources/repo/processes/general/BPMN2-SimpleRestartTest.bpmn2:188]: done

File: jbpm-services/jbpm-kie-services/src/test/resources/repo/processes/general/BPMN2-SimpleRestartWithObsoleteTest.bpmn2 [hunk 27]
```diff
@@ -0,0 +1,248 @@
...
+      <bpmn2:incoming>_D8CD8010-992B-447D-80D8-89E4B843A1C0</bpmn2:incoming>
+    </bpmn2:endEvent>
+    <bpmn2:parallelGateway id="_2C427968-15F3-4D6E-AA17-4AB10BAFF96C" gatewayDirection="Diverging">
+      <bpmn2:incoming>_85B212F7-1001-4FD7-9402-3814BCD42438</bpmn2:incoming>
+      <bpmn2:outgoing>_6E39142B-945E-43C2-B0B2-D27ED36D26A5</bpmn2:outgoing>
+      <bpmn2:outgoing>_3A02094A-DCAE-495A-8F89-246756403E63</bpmn2:outgoing>
+    </bpmn2:parallelGateway>
+    <bpmn2:endEvent id="_8AD18C56-2A0F-4A1B-AFAD-4BD505044EDD">
+      <bpmn2:incoming>_0B564073-D481-4768-A89E-90C19FD41459</bpmn2:incoming>
+      <bpmn2:terminateEventDefinition id="_aajkdJN-Eeq_IOLSelAS2A"/>
```
[the remark this answers is not in the document]
Comment (elguardian) [reply jbpm-services/jbpm-kie-services/src/test/resources/repo/processes/general/BPMN2-SimpleRestartWithObsoleteTest.bpmn2:145]: done

File: jbpm-services/jbpm-kie-services/src/test/resources/repo/processes/general/BPMN2-SimpleRestartWithObsoleteTest.bpmn2 [hunk 28]
```diff
@@ -0,0 +1,248 @@
...
+    </bpmn2:userTask>
+    <bpmn2:startEvent id="_58709295-E375-48ED-B835-BBCBD6C4F752">
+      <bpmn2:outgoing>_85B212F7-1001-4FD7-9402-3814BCD42438</bpmn2:outgoing>
+    </bpmn2:startEvent>
+    <bpmn2:endEvent id="_71B8E54A-CB9D-431D-9BD3-5C23C5371F40">
+      <bpmn2:incoming>_D8CD8010-992B-447D-80D8-89E4B843A1C0</bpmn2:incoming>
+    </bpmn2:endEvent>
+    <bpmn2:parallelGateway id="_2C427968-15F3-4D6E-AA17-4AB10BAFF96C" gatewayDirection="Diverging">
+      <bpmn2:incoming>_85B212F7-1001-4FD7-9402-3814BCD42438</bpmn2:incoming>
+      <bpmn2:outgoing>_6E39142B-945E-43C2-B0B2-D27ED36D26A5</bpmn2:outgoing>
```
[the remark this answers is not in the document]
Comment (elguardian) [reply jbpm-services/jbpm-kie-services/src/test/resources/repo/processes/general/BPMN2-SimpleRestartWithObsoleteTest.bpmn2:140]: nak

File: jbpm-flow/src/main/java/org/jbpm/workflow/instance/impl/NodeInstanceImpl.java [hunk 29]
```diff
@@ -78,12 +82,14 @@
     protected int slaCompliance = ProcessInstance.SLA_NA;
     protected Date slaDueDate;
     protected long slaTimerId = -1;
-    protected boolean aborted = false;
+    private boolean aborted = false;
+    transient protected CancelType cancelType;
```
[the remark this answers is not in the document]
Comment (elguardian) [reply jbpm-flow/src/main/java/org/jbpm/workflow/instance/impl/NodeInstanceImpl.java:86]: ok

File: jbpm-services/jbpm-kie-services/src/test/resources/repo/processes/general/BPMN2-SimpleRestartWithTimerTest.bpmn2 [hunk 30]
```diff
@@ -0,0 +1,352 @@
...
+    </bpmn2:endEvent>
+    <bpmn2:boundaryEvent id="_BAB8152A-00DC-436D-A4FD-0CC4D3DF3DB6" drools:boundaryca="true" drools:dockerinfo="58.0^74.0|" attachedToRef="_1A124FE6-9C2F-425E-807D-BB32FDD1304D">
+      <bpmn2:outgoing>_92EA16BB-8173-46A1-BCF5-1B7F1D629D38</bpmn2:outgoing>
+      <bpmn2:timerEventDefinition id="_XV3vqJN3Eeq_IOLSelAS2A">
+        <bpmn2:timeDuration xsi:type="bpmn2:tFormalExpression" id="_XV3vqZN3Eeq_IOLSelAS2A">PT2S</bpmn2:timeDuration>
+      </bpmn2:timerEventDefinition>
+    </bpmn2:boundaryEvent>
+  </bpmn2:process>
+  <bpmndi:BPMNDiagram id="_XV3vqpN3Eeq_IOLSelAS2A">
+    <bpmndi:BPMNPlane id="_XV3vq5N3Eeq_IOLSelAS2A" bpmnElement="test.test">
```
[the remark this answers is not in the document]
Comment (elguardian) [reply jbpm-services/jbpm-kie-services/src/test/resources/repo/processes/general/BPMN2-SimpleRestartWithTimerTest.bpmn2:230]: ok

File: jbpm-services/jbpm-kie-services/src/test/resources/repo/processes/general/BPMN2-SimpleRestartWithObsoleteTest.bpmn2 [hunk 31]
```diff
@@ -0,0 +1,248 @@
...
+      </bpmn2:dataInputAssociation>
+      <bpmn2:dataInputAssociation id="_aajkcJN-Eeq_IOLSelAS2A">
+        <bpmn2:targetRef>_2B0C2902-D947-43BB-8F03-1615434F65FF_SkippableInputX</bpmn2:targetRef>
+        <bpmn2:assignment id="_aajkcZN-Eeq_IOLSelAS2A">
+          <bpmn2:from xsi:type="bpmn2:tFormalExpression" id="_aajkcpN-Eeq_IOLSelAS2A"><![CDATA[false]]></bpmn2:from>
+          <bpmn2:to xsi:type="bpmn2:tFormalExpression" id="_aajkc5N-Eeq_IOLSelAS2A">_2B0C2902-D947-43BB-8F03-1615434F65FF_SkippableInputX</bpmn2:to>
+        </bpmn2:assignment>
+      </bpmn2:dataInputAssociation>
+      <bpmn2:potentialOwner id="a0002c84-9cf9-4194-ad87-e38db9d982cf">
+        <bpmn2:resourceAssignmentExpression id="_XV3Ii5N3Eeq_IOLSelAS2A">
```
[the remark this answers is not in the document]
Comment (elguardian) [reply jbpm-services/jbpm-kie-services/src/test/resources/repo/processes/general/BPMN2-SimpleRestartWithObsoleteTest.bpmn2:127]: nack

**Comment by elguardian** [comment]
> 1. Edge case: a new process is created when passing as nodeIds a string array with length 0, though obviously no node is started. I think it should throw an exception

not necesarilly. once a process is created we might trigger any node. (it would be exactly as creating a process)

> 2. Add some test passing correlation key.

this is done at integration level.

> 3. One question: From the same aborted process, I can create N new processes starting from the same nodes, is it expected?

yep it is possible.

File: jbpm-services/jbpm-kie-services/src/test/resources/repo/processes/general/BPMN2-SimpleRestartTest.bpmn2 [hunk 32]
```diff
@@ -0,0 +1,351 @@
...
+          <bpmn2:to xsi:type="bpmn2:tFormalExpression" id="_UVLvq4rNEeqCP7d-9hkcyw">_A1826972-90C6-4C66-A622-67F3EA5BD034_SkippableInputX</bpmn2:to>
+        </bpmn2:assignment>
+      </bpmn2:dataInputAssociation>
+      <bpmn2:dataOutputAssociation id="_UVLvrIrNEeqCP7d-9hkcyw">
+        <bpmn2:sourceRef>_A1826972-90C6-4C66-A622-67F3EA5BD034_ht2_outOutputX</bpmn2:sourceRef>
+        <bpmn2:targetRef>var_a</bpmn2:targetRef>
+      </bpmn2:dataOutputAssociation>
+      <bpmn2:potentialOwner id="_sjFE3NZxEeWhSJ3J-UY8fAB">
+        <bpmn2:resourceAssignmentExpression id="_sjFE3dZxEeWhSJ3J-UY8fAB">
+          <bpmn2:formalExpression id="_sjFE3tZxEeWhSJ3J-UY8fABº">john</bpmn2:formalExpression>
```
Comment (gmunozfe) [inline jbpm-services/jbpm-kie-services/src/test/resources/repo/processes/general/BPMN2-SimpleRestartTest.bpmn2:140]: take care of the last char
